[AMDGPU] Emit a waitcnt instruction after each memory instruction

jwanggit86 opened the pull request:
This patch introduces a new command-line option for clang, namely, amdgpu-precise-mem-op. When this option is specified, a waitcnt instruction is generated after each memory load/store instruction. The counter values are always 0, but which counters are involved depends on the memory instruction.

jwanggit86 committed:
[AMDGPU] Emit a waitcnt instruction after each memory instruction

This patch introduces a new command-line option for clang, namely,
amdgpu-precise-mem-op. When this option is specified, a waitcnt instruction
is generated after each memory load/store instruction. The counter values are
always 0, but which counters are involved depends on the memory instruction.

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2561,6 +2567,70 @@ bool SIMemoryLegalizer::expandAtomicCmpxchgOrRmw(const SIMemOpInfo &MOI,
+      AMDGPU::Waitcnt Wait;
+
+      if (!(Inst.getDesc().TSFlags & SIInstrFlags::maybeAtomic)) {
+        if (TII->isSMRD(Inst)) {          // scalar
```
This logic would need updating again for GFX12. It seems like it's duplicating a lot of knowledge which is already implemented in SIInsertWaitcnts.

jwanggit86 replied:
That's a valid point. However, even though it does similar work, I'd say the similarity to SIInsertWaitcnt is only to a certain extent. The differences include: (1) This is for a different purpose, i.e. to support the so-called "precise memory mode", in particular precise memory exceptions, for certain GPUs. (2) This feature is optional while SIInsertWaintcnt is not. (3) The counter values in SIInsertWaitcnt are precise, while in this features the counters are simply set to 0.
If performance is a concern, pls note that this feature is controlled by a command-line option which by default is off. The user has to explicitly give the option for it to work. We assume the user knows there's extra work for the compiler when the option is turned on.

jwanggit86 replied:
@t-tye Hi Tony, do you have any comments? Note that the function for gfx10 and 11 is empty for now. I want to get some feedback before going further.

Pierre-vh replied:
Can we have a shared helper, e.g. in `SIInstrInfo`  for both? It's a lot of logic to duplicate

> The counter values in SIInsertWaitcnt are precise, while in this features the counters are simply set to 0.
That could just be a boolean switch in a shared helper

arsenm replied:
Both the memory legalizer and SIInsertWaitcnts are required passes. This feature is not optional if requested. I do think this makes more sense to belong in SIInsertWaitcnts

t-tye commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -641,6 +644,9 @@ class SIMemoryLegalizer final : public MachineFunctionPass {
   bool expandAtomicCmpxchgOrRmw(const SIMemOpInfo &MOI,
                                 MachineBasicBlock::iterator &MI);
 
+  bool GFX9InsertWaitcntForPreciseMem(MachineFunction &MF);
```
Should these be combined with the expand* functions? They are supposed to do all that is necessary to "legalize" the opcodes to meet the memory model. And this inserting waitcnts is just another piece of that expansion.

Combining it can also avoid inserting multiple waitcnt for the same memory operation.

Combining it may be able to use the existing operation to ensure a memory operation is completed. I believe that operation should already be determining what kind of waitcnts should be inserted. If not, then I would consider generalizing it so it can be used by both the atomics expansion and the precise memory expansion.

It also keeps the operations in this class architecture neutral.

Pierre-vh replied:
Agreed, this should definitely be a virtual function such as `insertWaitcntForPreciseMem` and let the CacheControl implementation do what is needed. This is just emulating what `CacheControl` already does

jwanggit86 replied:
Sure. Code has been updated as suggested.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPU.td:
```diff
@@ -167,6 +167,10 @@ def FeatureCuMode : SubtargetFeature<"cumode",
   "Enable CU wavefront execution mode"
 >;
 
+def FeaturePreciseMemory
```
I'm not a fan of using a feature for this, I think we should have a backend CL option instead. After all this isn't an architecture feature but just a behavior change for SIMemoryLegalizer.

jwanggit86 replied:
End user needs this, so a clang command-line option has been added. There was a lot discussion on this. I'll forward you the email.

Pierre-vh replied:
Understood :) 
Can you remove the `amdgpu` prefix from the option? All target features are already specific to AMDGPU, e.g. xnack, sramecc, etc. so you can just use something like `precise-memory`

jwanggit86 replied:
The name was the result of some discussions last year. I've forwarded you the email.

arsenm replied:
The subtarget feature prefix should be removed. The subtarget feature name is not the user facing component of this, and no user should be directly using it

jwanggit86 replied:
As it is, we have a clang command-line option "-mamdgpu-precise-memory-op". When specified, "+amdgpu-precise-memory-op" would be passed to the backend as a target feature. So if the backend uses a different name for this feature, e.g., dropping the 'amdgpu' prefix as you suggested, the target feature passed by clang will be considered invalid and silently ignored.

In short, my understanding is that if the prefix is dropped in the backend (in `AMDGPU.td`), then the clang command-line option also has to drop it. If you know how to keep it in clang but drop it in the backend, pls let me know.

jwanggit86 replied:
Thanks for the suggestion. It looks it would work. However, I'm not sure the extra overhead is worth it just for having a shorter name in the backend. After all this is not going to be a commonly used option, and the name in the backend is not exposed to end user.

Pierre-vh replied:
The extra overhead is just 3 lines in `clang/lib/Driver/ToolChains/AMDGPU.cpp`, it's negligible.
We don't have any target feature that starts with `amdgpu` so it's out of place IMO, it should really drop the `amdgpu-` prefix.

jwanggit86 replied:
The function `hasFlag()` actually appears to be pretty expensive, and it's going to be called every time, regardless of whether the flag is used or not.

Let's get Matt's opinion on this. @arsenm What do you think?

Pierre-vh replied:
It's only called once per run by the driver, yes

We already do this for wavefrontsize64, and pretty much all other options that aren't automatically handled. We even call `hasFlag` twice on the same option in some cases instead of saving the result somewhere, so I doubt `hasFlag` can be considered expensive enough to justify not dropping `amdgpu-` from the feature name

jwanggit86 replied:
@Pierre-vh With the suggested change, the func `getAMDGPUTargetFeatures` looks something like the following:
```
void amdgpu::getAMDGPUTargetFeatures(...) {
...
  if (Args.hasFlag(options::OPT_mwavefrontsize64,
                   options::OPT_mno_wavefrontsize64, false))
    Features.push_back("+wavefrontsize64");

  if (Args.hasFlag(options::OPT_mamdgpu_precise_memory_op,
                   options::OPT_mno_amdgpu_precise_memory_op, false)) {
    Features.push_back("+precise-memory");
  }
  handleTargetFeaturesGroup(D, Triple, Args, Features,
                            options::OPT_m_amdgpu_Features_Group);
}

However, `handleTargetFeaturesGroup` does not seem to care whether an Arg is claimed or not. It will process every Arg, and we end up with the following:
`"-target-feature" "+precise-memory" "-target-feature" "+amdgpu-precise-memory-op"`

jwanggit86 replied:
A possible fix is to check `hasFlag` after calling `handleTargetFeaturesGroup`, and if the check is true, replace the string `"+amdgpu-precise-memory-op"` in the `Features` vector with `"+precise-memory"`, as follows:
```
  handleTargetFeaturesGroup(D, Triple, Args, Features,
                            options::OPT_m_amdgpu_Features_Group);

  if (Args.hasFlag(options::OPT_mamdgpu_precise_memory_op,
                   options::OPT_mno_amdgpu_precise_memory_op, false)) {
    for (auto It = Features.begin(); It != Features.end(); ++It) {
      if (*It == "+amdgpu-precise-memory-op")
        *It = StringRef("+precise-memory");
    }
```
@Pierre-vh Your thoughts?

Pierre-vh replied:
Just remove `m_amdgpu_Features_Group` from your option's `SimpleMFlag`, follow the same pattern as `wavefrontsize64`

jwanggit86 replied:
Thanks! This works. Code has been updated. Pls take a look.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -17,13 +17,16 @@
 #include "llvm/Support/AtomicOrdering.h"
 #include "llvm/TargetParser/TargetParser.h"
 
+#include <iostream>
```
I suspect this is a debug leftover but: always use `dbgs` to print, it's easier and faster to build than `iostream` :)

jwanggit86 replied:
Removed.

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_all.ll:
```diff
@@ -0,0 +1,199 @@
+; GFX9-FLATSCR:        scratch_load_dword {{v[0-9]+}}, {{v[0-9]+}}, off
+; GFX9-FLATSCR-NEXT:   s_waitcnt vmcnt(0) lgkmcnt(0)
+define amdgpu_vs float @vs_main(i32 %idx) {
+  %v1 = extractelement <81 x float> <float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef, float undef ...
```
what is that test for/why does it need such a big vector?
if it's to force a spill to stack I think you can do that by playing with the number of available v/sgprs, IIRC there is an attribute or CL opt for that

jwanggit86 replied:
The function is copied from another test case. The purpose was to generate one or more particular memory instructions. In this particular case, "scratch_load" and "scratch_store".

arsenm replied:
if you just want scratch load and store, it's much simpler to just directly write a simple test that uses that.

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_all.ll:
```diff
@@ -0,0 +1,199 @@
+; Testing the -amdgpu-precise-memory-op option
```
Please generate the test using `update_llc_test_checks`, much easier to update if/when things change.

Also I think you don't need `-verify-machineinstrs`. It's expensive and runs anyway when EXPENSIVE_CHECKS is on.

arsenm commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2561,6 +2567,70 @@ bool SIMemoryLegalizer::expandAtomicCmpxchgOrRmw(const SIMemOpInfo &MOI,
   return Changed;
 }
 
+bool SIMemoryLegalizer::GFX9InsertWaitcntForPreciseMem(MachineFunction &MF) {
```
can you just make this happen as a consequence of the existing flow rather than using an entirely separate pass over the function?

jwanggit86 replied:
Yes. Code has been rewritten.

arsenm commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2561,6 +2567,70 @@ bool SIMemoryLegalizer::expandAtomicCmpxchgOrRmw(const SIMemOpInfo &MOI,
+    for (auto MI = MBB.begin(); MI != MBB.end();) {
+      MachineInstr &Inst = *MI;
+      ++MI;
+      if (Inst.mayLoadOrStore() == false)
```
Don't use == false

jwanggit86 replied:
Ok.

jwanggit86 committed:
Combined insertions of waitcnt with existing SIMemoryLegalizer code.

jwanggit86 commented:
@t-tye Code has been updated based on your feedback. Pls take a look.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -605,12 +606,197 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
                                       bool IsNonTemporal) const override;
 };
 
+class SIPreciseMemorySupport {
```
Why does it need to be a separate class hierarchy?
It could just be part of CacheControl, and the functions can be named `handlePreciseMemoryAtomic/NonAtomic` ?
That would avoid a lot of boilerplate.

jwanggit86 replied:
I don't have strong objection to merging with CacheControl, but would like to get Tony's opinion. @t-tye Tony, what do you think?

t-tye replied:
My initial thought had been that this would be part of the existing cache control functions. It seems it is the same kind of waitcnt as needs to be inserted after a store release. That also requires the right waitcnt to be generated according to the kind of memory instruction.

jwanggit86 replied:
Ok, will merge with CacheControl.

jwanggit86 replied:
Merged with SICacheControl.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -605,12 +606,197 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
+std::unique_ptr<SIPreciseMemorySupport>
+SIPreciseMemorySupport::create(const GCNSubtarget &ST) {
+  GCNSubtarget::Generation Generation = ST.getGeneration();
+  if (Generation < AMDGPUSubtarget::GFX10)
```
GFX12 is missing

jwanggit86 replied:
Not sure this is required for GFX12. @t-tye Tony, is this required for GFX12? We didn't discuss this for GFX12.

t-tye replied:
Is there a reason that this functionality should not be available for any target? It is true it is only particularly useful for targets that have no precise memory operations hardware support, but the basic idea is meaningful for any target.

jwanggit86 replied:
Pls take a look at the updated code. I'll be working on gfx12 in the meantime.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -605,12 +606,197 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
+}
+
+bool SIGfx10And11PreciseMemorySupport ::handleAtomic(
+    MachineBasicBlock::iterator &MI, bool ret) {
```
Use something more descriptive than `ret` + use CamelCase, e.g. `IsReturningAtomic`

jwanggit86 replied:
Like the existing functions, the return value indicates whether there's any change to the basic block, e.g., inserting a new instruction.

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,362 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 -mattr=+amdgpu-precise-memory-op < %s | FileCheck %s -check-prefixes=GFX9
+; RUN: llc -mtriple=amdgcn -mcpu=gfx90a -mattr=+amdgpu-precise-memory-op < %s | FileCheck %s -check-prefixes=GFX90A
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -mattr=+amdgpu-precise-memory-op < %s | FileCheck %s -check-prefixes=GFX10
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx900 -mattr=-flat-for-global,+enable-flat-scratch,+amdgpu-precise-memory-op -amdgpu-use-divergent-register-indexing < %s | FileCheck --check-prefixes=GFX9-FLATSCR %s
```
gfx11 and 12 tests

jwanggit86 replied:
Tests for gfx11 added. Gfx12 is on-going.

Pierre-vh reviewed (changes requested):
When you made changes, you can click the "Re-request review" icon next to reviewers to put it back in the review queues :)

jayfoad commented:
> This logic would need updating again for GFX12. It seems like it's duplicating a lot of knowledge which is already implemented in SIInsertWaitcnts.

Just to demonstrate, you could implement this feature in SIInsertWaitcnts for **all** supported architectures with something like:
```diff
diff --git a/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp b/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp
index 6ecb1c8bf6e1..910cd094f8f2 100644
--- a/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp
+++ b/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp
@@ -2299,6 +2299,12 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
 
     updateEventWaitcntAfter(Inst, &ScoreBrackets);
 
+    AMDGPU::Waitcnt Wait =
+        AMDGPU::Waitcnt::allZeroExceptVsCnt(ST->hasExtendedWaitCounts());
+    ScoreBrackets.simplifyWaitcnt(Wait);
+    Modified |= generateWaitcnt(Wait, std::next(Inst.getIterator()), Block,
+                                ScoreBrackets, /*OldWaitcntInstr=*/nullptr);
+
 #if 0 // TODO: implement resource type check controlled by options with ub = LB.
     // If this instruction generates a S_SETVSKIP because it is an
     // indexed resource, and we are on Tahiti, then it will also force
```
Handling VSCNT/STORECNT correctly is a little more complicated but not much.

jwanggit86 committed:
Merge code for precise mem with the existing SICacheControl classes.

jwanggit86 committed:
Add support for GFX12.

Pierre-vh reviewed (commented):
I also agree with Jay, can't this go in InsertWaitCnt? Why does it have to go in SIMemoryLegalizer instead?

If it has to stay here, fine, but is it possible to merge some code with SIInsertWaitCnt in a common helper somewhere?

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -603,14 +626,69 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
+  bool
+  handleNonAtomicForPreciseMemory(MachineBasicBlock::iterator &MI) override;
+  bool handleAtomicForPreciseMemory(MachineBasicBlock::iterator &MI,
+                                    bool ret) override;
```
Changed "ret" to "IsAtomicWithRet".

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -603,14 +626,69 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
+                                    bool ret) override;
+};
+
+#if 0
```
Done.

jwanggit86 replied:
Regarding the question about SIInsertWaitcnt, initially the code was indeed put there (see PR [68932](https://github.com/llvm/llvm-project/pull/68932)). After a discussion with @t-tye and others, it was decided to do it in SIMemoryLegalizer. I think the main reason is that functionality-wise this new feather is more closely related to the Mem Legalizer. Tony can explain this better. Also, the Mem Legalizer, like SIInsertWaitcnt, goes through each instruction already.
Based on the implementation experience, I'd say another reason that supports putting it in the Mem Legalizer is that there's already a class hierarchy for the different ISAs, making it easier to implement ISA-specific behaviors.

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -603,14 +626,69 @@ class SIGfx12CacheControl : public SIGfx11CacheControl {
   std::unique_ptr<SICacheControl> CC = nullptr;
 
+  /// Precise Memory support.
+  bool PM = false;
```
Changed "PM" to "PrecMem".

jwanggit86 committed:
Some small changes based on code review.

jwanggit86 committed:
Code formatting.

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2378,6 +2456,221 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+    BuildMI(MBB, ++MI, DebugLoc(), TII->get(AMDGPU::S_WAIT_LOADCNT)).addImm(0);
+  } else {
+    BuildMI(MBB, ++MI, DebugLoc(), TII->get(AMDGPU::S_WAIT_STORECNT)).addImm(0);
+  }
```
Done.

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2378,6 +2456,221 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+        WaitType = AMDGPU::S_WAIT_STORECNT_DSCNT;
+      } else {
+        WaitType = AMDGPU::S_WAIT_DSCNT;
+      }
```
Done.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -355,6 +356,18 @@ class SICacheControl {
+
+  virtual bool
+  handleNonAtomicForPreciseMemory(MachineBasicBlock::iterator &MI) = 0;
+  /// Handles atomic instruction \p MI with \p IsAtomicWithRet indicating
+  /// whether \p MI returns a result.
```
IMO the comment isn't needed, I think the function's name describes it well enough (and the comment just before adds enough context), the comment just rephrases the function + argument names

jwanggit86 replied:
Adding the comments is mainly to be consistent with the existing code. Here the comment points out that the 2nd parameter is a property of the first parameter (MI), which is not absolutely clear by just looking at the code.

Pierre-vh reviewed (changes requested):
Did you try to move this to SIInsertWaitCnt, as suggested?

jwanggit86 committed:
Change the option from amdgpu-precise-memory-op to precise-memory
for the backend.

jwanggit86 commented:
@Pierre-vh Any further comments?

Pierre-vh reviewed (approved):
LGTM, but wait for @t-tye or @jayfoad to approve as well

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -355,6 +356,18 @@ class SICacheControl {
+  handleNonAtomicForPreciseMemory(MachineBasicBlock::iterator &MI) = 0;
+  /// Handles atomic instruction \p MI with \p IsAtomicWithRet indicating
+  /// whether \p MI returns a result.
+  virtual bool handleAtomicForPreciseMemory(MachineBasicBlock::iterator &MI,
```
This function is never even called.

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2378,6 +2409,215 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+      LoadAndStore = true;
+    } else if (Inst.mayLoad()) { // vector load
+      if (TII->isVMEM(Inst))     // VMEM load
+        WaitType = AMDGPU::S_WAIT_LOADCNT;
```
What about SAMPLECNT and BVHCNT?

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2378,6 +2409,215 @@ bool SIGfx12CacheControl::enableVolatileAndOrNonTemporal(
+
+  MachineInstr &Inst = *MI;
+  unsigned WaitType = 0;
+  // For some vector instructions, mayLoad() and mayStore() can be both true.
```
What kind of (non-atomic) instructions is this supposed to handle?

jayfoad reviewed (changes requested):
I've added _some_ inline comments, but really I don't want to spend the time to review this properly (or maintain it, or extend it for new architectures in future). All this logic already exists in SIInsertWaitcnts. Duplicating it here is not a good design.

t-tye commented:
I am not clear why new functions need to be added for this, as I think there are existing functions that already do this.

jwanggit86 committed:
Move implementation from SIMemoryLegalizer to SIInsertWaitcnts.

jwanggit86 commented:
@jayfoad After trying the patch you provided above, it appears that this feature can indeed be done in SIInsertWaitcnt instead of SIMemoryLegalizer. Code has been updated accordingly. Pls take a look.

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2326,6 +2326,20 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
 
+    if (ST->isPreciseMemoryEnabled()) {
+      AMDGPU::Waitcnt Wait;
+      if (WCG == &WCGPreGFX12)
```
Use `ST->hasExtendedWaitCounts()` instead of checking the pointer?

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2326,6 +2326,20 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
+    if (ST->isPreciseMemoryEnabled()) {
+      AMDGPU::Waitcnt Wait;
+      if (WCG == &WCGPreGFX12)
+        Wait = AMDGPU::Waitcnt(0, 0, 0, 0);
```
I was looking at https://github.com/ROCm/ROCm-CompilerSupport/issues/66 and it made me wonder, why do we have to emit all zeroes instead of just emitting what's in `ScoreBrackets`? Is there an advantage?

I'm wondering if this should just emit `ScoreBrackets`, then `+precise-memory` + `-amdgpu-waitcnt-forcezero` need to be used together achieve the behavior we have here?

jwanggit86 replied:
The option `amdgpu-waitcnt-forcezero` appears to force an s_waitcnt(0) before EVERY instruction.
With this example,
```
define void @syncscope_workgroup_nortn(ptr %addr, float %val) {
  %res = atomicrmw fadd ptr %addr, float %val syncscope("workgroup") seq_cst
  ret void
}
```
The .s is as follows:
```
; %bb.0:
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        flat_load_dword v4, v[0:1]
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        s_mov_b64 s[4:5], 0
.LBB0_1:                                ; %atomicrmw.start
                                        ; =>This Inner Loop Header: Depth=1
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        v_add_f32_e32 v3, v4, v2
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        flat_atomic_cmpswap v3, v[0:1], v[3:4] glc
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        v_cmp_eq_u32_e32 vcc, v3, v4
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        s_or_b64 s[4:5], vcc, s[4:5]
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        v_mov_b32_e32 v4, v3
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        s_andn2_b64 exec, exec, s[4:5]
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        s_cbranch_execnz .LBB0_1
; %bb.2:                                ; %atomicrmw.end
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        s_or_b64 exec, exec, s[4:5]
        s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
        s_setpc_b64 s[30:31]
```
So, it doesn't seem to be helpful for either this issue or [issue#66](https://github.com/ROCm/ROCm-CompilerSupport/issues/66).

arsenm reviewed (changes requested):
Outstanding comments

jwanggit86 committed:
Minor code change.

arsenm commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,577 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 -mattr=+precise-memory < %s | FileCheck %s -check-prefixes=GFX9
+; RUN: llc -mtriple=amdgcn -mcpu=gfx90a -mattr=+precise-memory < %s | FileCheck %s -check-prefixes=GFX90A
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1010 -mattr=+precise-memory < %s | FileCheck %s -check-prefixes=GFX10
+; RUN: llc -mtriple=amdgcn-- -mcpu=gfx900 -mattr=-flat-for-global,+enable-flat-scratch,+precise-memory -amdgpu-use-divergent-register-indexing < %s | FileCheck --check-prefixes=GFX9-FLATSCR %s
```
Don't use -flat-for-global, the only reason we really have that is to avoid updating shared tests which doesn't apply if you're just generating these

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,577 @@
+; GFX9-FLATSCR-NEXT:   s_waitcnt vmcnt(0)
+; GFX9-FLATSCR:        scratch_load_dword {{v[0-9]+}}, {{v[0-9]+}}, off
+; GFX9-FLATSCR-NEXT:   s_waitcnt vmcnt(0)
+define amdgpu_vs float @vs_main(i32 %idx) {
```
This is a convoluted way of getting stack access. Just directly load/store from an addrspace(5) pointer

jwanggit86 replied:
Removed this testcase and added a new one that uses addrspace(5).

arsenm commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,577 @@
+  %v.0 = load i32, ptr addrspace(3) %ptr, align 8
+  %v.1 = load i32, ptr addrspace(3) %ptr.gep.1
+
+  %r.0 = insertelement <2 x i32> undef, i32 %v.0, i32 0
```
Use poison instead of undef throughout

jwanggit86 replied:
Done.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2326,6 +2326,20 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
 #endif
 
+    if (ST->isPreciseMemoryEnabled() && Inst.mayLoadOrStore()) {
+      AMDGPU::Waitcnt Wait;
+      if (ST->hasExtendedWaitCounts())
+        Wait = AMDGPU::Waitcnt(0, 0, 0, 0, 0, 0, 0);
+      else
+        Wait = AMDGPU::Waitcnt(0, 0, 0, 0);
+
+      if (!Inst.mayStore())
+        Wait.StoreCnt = ~0u;
```
```suggestion
      AMDGPU::Waitcnt Wait = WCG->getAllZeroWaitcnt(Inst.mayStore());
```
However, as a general rule:
- loads and atomics-with-return update LOADcnt
- stores and atomics-without-return update STOREcnt

so it might be more accurate to use the condition `Inst.mayStore() && !SIInstrInfo::isAtomicRet(Inst)`.

Please make sure you have tests for atomics with and without return.

jwanggit86 replied:
Code updated as suggested. Testfile includes case for both atomic-with-ret and atomic-no-ret. However, for the following case, even though `ds_add_u32` is atomic-no-ret, the Waitcnt for StoreCnt is set to ~0u after the call of `ScoreBrackets.simplifyWaitcnt(Wait)`. Therefore, no s_waitcnt for the StoreCnt is generated after the `ds_add_u32`.
```
define amdgpu_kernel void @atomic_add_local(ptr addrspace(3) %local) {
   %unused = atomicrmw volatile add ptr addrspace(3) %local, i32 5 seq_cst
   ret void
}
```
The code for GFX1100 is:
```
; GFX11:         ds_add_u32 v0, v1
; GFX11-NEXT:    s_waitcnt lgkmcnt(0)
; GFX11-NEXT:    buffer_gl0_inv
```
Pls let me know if this looks correct.

arsenm replied:
gfx11 doesn't have the separate load/store counters, that's gfx11. I would expect this to only use lgkmcnt here

jayfoad replied:
GFX10 introduced a separate counter for **VMEM** stores with the name VScnt. GFX12 just renamed it to STOREcnt. No architecture has a separate store counter for DS or SMEM. So `ds_add_u32 v0, v1` followed by `s_waitcnt lgkmcnt(0)` (pre-GFX12) or `s_wait_dscnt 0` (GFX12) is fine .

jwanggit86 replied:
Thanks! Is there anything else that should be addressed?

jwanggit86 committed:
Use getAllZeroWaitcnt() when creating the Wait obj. Some changes
to the test file.

jwanggit86 committed:
Use update_llc_test_checks.py on insert_waitcnt_for_precise_memory.ll.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,618 @@
+; from atomicrmw-expand.ll
+; covers flat_load, flat_atomic (atomic with return)
+;
+; GFX90A-LABEL: syncscope_workgroup_nortn:
```
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,1413 @@
+; GFX12-NEXT:    global_store_b16 v[2:3], v0, off
+; GFX12-NEXT:    s_wait_storecnt 0x0
+; GFX12-NEXT:    s_setpc_b64 s[30:31]
+  %val = load bfloat, ptr addrspace(1) %in
+  store bfloat %val, ptr addrspace(1) %out
```
Using bfloat as your general test type is a bit weird. use i32?

jwanggit86 replied:
This testcase is removed. See below.

arsenm commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,1413 @@
+; GFX9:       ; %bb.0:
+; GFX9-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX9-NEXT:    global_load_ushort v0, v[0:1], off
+; GFX9-NEXT:    s_waitcnt vmcnt(0)
```
I'm not sure tests like this really show the effect of the flag. I bet you would end up with the same wait decisions without it. Can you verify there's a test diff with/without the flag, and add some intermediate operation padding that would cause deferred wait placement?

jwanggit86 replied:
Good catch. This testcase (test_load_store()) and another one, test_load_store_as5 have been replaced with `tail_call_byval_align16` which generates buffer_load/store and scratch_load/store.

jwanggit86 committed:
Replace testcases test_load_store() and test_load_store_as5() with
tail_call_byval_align16() to cover buffer_load/store, and
scratch_load/store.

jwanggit86 commented:
@jayfoad @arsenm Any other comments?

jayfoad commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2594,12 +2594,10 @@ bool SIMemoryLegalizer::expandAtomicCmpxchgOrRmw(const SIMemOpInfo &MOI,
-                                                   SIMemOp::STORE,
-                                MOI.getIsCrossAddressSpaceOrdering(),
-                                Position::AFTER);
+      Changed |=
```
Remove this.

jayfoad commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,1406 @@
+; GFX9-NEXT:    v_mov_b32_e32 v0, s0
+; GFX9-NEXT:    ds_add_u32 v0, v1
+; GFX9-NEXT:    s_waitcnt lgkmcnt(0)
+; GFX9-NEXT:    s_waitcnt lgkmcnt(0)
```
Why do we get two s_waitcnts? This looks wrong.

jayfoad commented on llvm/test/CodeGen/AMDGPU/insert_waitcnt_for_precise_memory.ll:
```diff
@@ -0,0 +1,1406 @@
+}
+
+; from atomic_load_add.ll
+; covers s_load, ds_add (atomic without return)
```
Can you add a test for DS atomic with return, e.g. ds_add_rtn?

jwanggit86 replied:
Done.

jwanggit86 committed:
Check if mem instruciton is already immediately followed by a
waitcnt instruction. If so, do not insert another waitcnt. Also
add a testcase that has ds_add_rtn. Formatting change made to
SIMemoryLegalizer.cpp is reverted.

arsenm commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2327,11 +2327,15 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
-      ScoreBrackets.simplifyWaitcnt(Wait);
-      Modified |= generateWaitcnt(Wait, std::next(Inst.getIterator()), Block,
-                                  ScoreBrackets, /*OldWaitcntInstr=*/nullptr);
+      Iter++;
```
```suggestion
      ++Iter;
```

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2327,11 +2327,15 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
+        Modified |= generateWaitcnt(Wait, std::next(Inst.getIterator()), Block,
+                                    ScoreBrackets, /*OldWaitcntInstr=*/nullptr);
+      }
+      Iter--;
```
You shouldn't need to manipulate the iterator here. Do you just need to get rid of the std::next, or use Iter directly above?

jwanggit86 replied:
When I see a load/store instruction, I need to check if the next one is a waitcnt. A waitcnt is only inserted if the next one is not a waitcnt. Thus the Iter++ and Iter--.

arsenm replied:
Something is wrong, you shouldn't need to check for neighbors?

jwanggit86 committed:
Change iterator update operator from post-inc to pre-inc.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2326,6 +2326,18 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
 
+    if (ST->isPreciseMemoryEnabled() && Inst.mayLoadOrStore()) {
+      ++Iter;
+      if (!isWaitInstr(*Iter)) {
```
This isn't right. This will prevent adding youtr waitcnt if *any* waitcnt follows, even a wait for a different counter.

Normally `generateWaitcnt` should combine and remove redundant waitcnts like the ones in your test. It looks like the problem is that this does not normally happen at the end of a basic block. I have tried fixing that in #87539. If that patch is committed, you should be able to rebase this patch and the duplicate waitcnts will get removed.

jayfoad replied:
#87539 has been merged. You can merge-or-rebase main into your branch and remove this `isWaitInstr` check.

jwanggit86 committed:
With #87539, previous commit that checks for the instruction immediately
after a load/store is not necessary.

jayfoad commented:
Can you add at least one test for a VMEM (flat or scratch or global or buffer or image) atomic without return? That should use vscnt on GFX10.

Apart from that the SIInsertWaitcnts.cpp and tests look good to me. I have not reviewed the clang parts but it looks like @Pierre-vh approved them previously?

jwanggit86 committed:
Add testcase that covers flat_atomic_swap, an atomic without return.

jwanggit86 commented:
Added a testcase that has flat_atomic_swap, which is an atomic without return.

jwanggit86 commented:
@jayfoad Do you have any more comments?

jayfoad commented:
No further comments.
